embed/serve.go

```diff
@@ -248,21 +250,74 @@ func (sctx *serveCtx) createMux(gwmux *gw.ServeMux, handler http.Handler) *http.
 	return httpmux
 }
 
-// wraps HTTP multiplexer to mute requests to /v3beta
-// TODO: deprecate this in 3.5 release
-func wrapMux(mux *http.ServeMux) http.Handler { return &v3Mutator{mux: mux} }
+// wrapMux wraps HTTP multiplexer:
+// - mutate gRPC gateway request paths
+// - check hostname whitelist
+// client HTTP requests goes here first
+func wrapMux(s *etcdserver.EtcdServer, mux *http.ServeMux) http.Handler {
+	return &httpWrapper{
+		s:   s,
+		mux: mux,
+	}
+}
 
-type v3Mutator struct {
+type httpWrapper struct {
+	s   *etcdserver.EtcdServer
 	mux *http.ServeMux
 }
 
-func (m *v3Mutator) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+func (m *httpWrapper) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	// redirect for backward compatibilities
 	if req != nil && req.URL != nil && strings.HasPrefix(req.URL.Path, "/v3beta/") {
 		req.URL.Path = strings.Replace(req.URL.Path, "/v3beta/", "/v3/", 1)
 	}
+
+	// enforce client origin (CVE-2018-5702)
+	// https://bugs.chromium.org/p/project-zero/issues/detail?id=1447#c2
+	if req != nil {
+		//  1. If client connection is secured via HTTPS or auth is enabled,
+		//     then allow any hostnames.
+		//  2. If client connection is not secured or auth is not enabled,
+		//     then allow only hostnames of loopback address (e.g. localhost)
+		//     or the hostname exists in given host whitelist.
+		host := netutil.GetHost(req.Host)
+
+		// allow if loopback address or "*"
+		ok := netutil.IsLoopback(host) || m.s.IsHostWhitelisted("*")
+		if !ok {
+			// connection is not secure, or auth is not enabled
+			if req.TLS == nil || !m.s.AuthStore().IsAuthEnabled() {
+				// allow if host is in whitelist
+				ok = m.s.IsHostWhitelisted(host)
```

Comment:
We still want to enforce a Host whitelist even if auth is enabled, right? I'd just recommend

```go
if !m.s.IsHostWhitelisted("*") && !m.s.IsHostWhitelisted(host) {
    // error
}
```